jerry-core/ecma/builtin-objects/ecma-builtin-helpers-date.cpp

```diff
@@ -733,25 +734,58 @@ ecma_date_make_day (ecma_number_t year, /**< year value */
     return ecma_number_make_nan ();
   }
 
+  /* 2., 3., 4. */
   ecma_number_t y = ecma_number_trunc (year);
   ecma_number_t m = ecma_number_trunc (month);
   ecma_number_t dt = ecma_number_trunc (date);
+  /* 5. */
   ecma_number_t ym = y + (ecma_number_t) floor (m / 12);
+  /* 6. */
   ecma_number_t mn = (ecma_number_t) fmod (m, 12);
   mn = (mn < 0) ? 12 + mn : mn;
-  ecma_number_t time = ecma_date_time_from_year (ym);
 
-  JERRY_ASSERT (ecma_date_year_from_time (time) == ym);
+  /* 7. We need to find the time where: YearFromTime(time) == ym
+   *    and MonthFromTime(time) == mn and DateFromTime(time) == 1.
+   *    If this is not possible return NaN.
+   */
+  ecma_number_t time = ecma_date_time_from_year (ym);
+  ecma_number_t step = floor ((ecma_number_t) (365) / 2);
+  ecma_number_t delta = step;
 
-  while (ecma_date_month_from_time (time) < mn)
+  if (ecma_date_year_from_time (time) == ym)
   {
-    time += ECMA_DATE_MS_PER_DAY;
-  }
+    /* Binary search for the closest day */
+    while (delta)
+    {
+      if (ecma_date_month_from_time (time + step * ECMA_DATE_MS_PER_DAY) < mn)
+      {
+        step += delta;
+      }
+      else
+      {
+        step -= delta;
+      }
+      delta = floor (delta / 2);
+    }
+
+    if (ecma_date_month_from_time (time + step) < mn)
+    {
+      time += step * ECMA_DATE_MS_PER_DAY;
+    }
+
+    while (ecma_date_month_from_time (time) < mn)
```

Comment:
No, because the binary search has to use integer days so the approximation is given by the binary search needs some fine tuning here.